sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageStreamSource.java

```diff
@@ -219,7 +232,15 @@ private synchronized boolean readNextRecord() throws IOException {
         }
 
         fractionConsumedFromPreviousResponse = fractionConsumedFromCurrentResponse;
-        ReadRowsResponse currentResponse = responseIterator.next();
+        ReadRowsResponse currentResponse;
+        Stopwatch stopwatch = Stopwatch.createStarted();
```

Comment:
+1 to adding an options to allow disabling this.

In the past we have avoided using tools which grab the system clock frequently as this system call can be quite slow and add performance impact. Especially if its done frequently. In this case however, it looks like the system clock may only be called infrequently. I.e. whenever an RPC is made (doing it or every element would be an issue) In which case its fine to do so.

Instead we have used the state sampler classes to do this. This uses a separate thread to count which periodically wakes up and checks what "state" other threads are executing under. Which would . 

However, there is no user metric or sdk level API to time blocks of code using the state sampler.

We only used it for the execution time metrics. To calculate the time spend in the process, startBundle and finishBundle methods of each ParDos. Here it is for reference.
https://github.com/apache/beam/pull/7676

Reply:
I can certainly add options to disable the collection of these metrics if we establish that this is the right thing to do.

The ReadRows API is a server streaming API which, on the server side, tries to send ReadRowsResponse messages as fast as the client can accept them. Each response message contains up to 100 MiB of encoded data -- in the extreme case, this is one row per response, but in the common case this is a batch of rows, usually 1024. The timer is read once -- or, I guess, twice -- per response message.

What I'd like to establish with these metrics is whether the existing client-side read-ahead is sufficient in the common case to prevent (a large amount of) blocking while reading from a stream, or whether we're spending time waiting for I/O on this thread in addition to doing Avro decoding, etc.

Reply:
It's probably fine to not use an option, since it occurs infrequently. Though it may be a good idea to add an options as a precaution. Making it opt in. We can turn it on by default later.

It may be a lot of effort to do this with the state sampler. So the approach you have is probably best to start with.

Other approaches would be to run the job through some sort of profiler. Though I am not sure how to do that today with any of our runners. Possibly it could be setup with direct runner. I think that would help you to get plenty of useful info on the problem as well. May be worth asking on the dev list about that.